src/backend/dx11/src/internal.rs

```diff
@@ -691,4 +834,124 @@ impl Internal {
+            context.PSSetConstantBuffers(0, 1, [self.internal_buffer.as_raw()].as_ptr());
+        }
+
+        let subpass = &cache.render_pass.subpasses[cache.current_subpass];
+
+        for clear in clears {
+            let clear = clear.borrow();
+
+            let _scope = debug_scope!(context, "{:?}", clear);
+
+            match *clear {
+                command::AttachmentClear::Color { index, value } => {
+                    self.update_clear_color(context, value);
+
+                    let attachment = {
+                        let rtv_id = subpass.color_attachments[index];
+                        &cache.framebuffer.attachments[rtv_id.0]
+                    };
+
+                    unsafe {
+                        context.OMSetRenderTargets(1, [attachment.rtv_handle.clone().unwrap().as_raw()].as_ptr(), ptr::null_mut());
+                    }
+
+                    match value {
+                        command::ClearColor::Float(_) => {
+                            unsafe {
+                                context.PSSetShader(self.ps_partial_clear_float.as_raw(), ptr::null_mut(), 0);
+                            }
+                        }
+                        command::ClearColor::Uint(_) => {
+                            unsafe {
+                                context.PSSetShader(self.ps_partial_clear_uint.as_raw(), ptr::null_mut(), 0);
+                            }
+                        }
+                        command::ClearColor::Int(_) => {
+                            unsafe {
+                                context.PSSetShader(self.ps_partial_clear_int.as_raw(), ptr::null_mut(), 0);
+                            }
+                        }
+                    }
+
+                    for clear_rect in &clear_rects {
+                        let viewport = conv::map_viewport(&Viewport {
+                            rect: clear_rect.rect,
+                            depth: 0f32..1f32,
+                        });
+
+                        debug_marker!(context, "{:?}", clear_rect.rect);
+
+                        unsafe {
+                            context.RSSetViewports(1, [viewport].as_ptr());
+                            context.Draw(3, 0);
+                        }
+                    }
+                }
+                command::AttachmentClear::DepthStencil { depth, stencil } => {
+                    self.update_clear_depth_stencil(context, depth, stencil);
+
+                    let attachment = {
+                        let dsv_id = subpass.depth_stencil_attachment.unwrap();
+                        &cache.framebuffer.attachments[dsv_id.0]
+                    };
+
+                    unsafe {
+                        match (depth, stencil) {
+                            (Some(_), Some(stencil)) => {
+                                context.OMSetDepthStencilState(self.partial_clear_depth_stencil_state.as_raw(), stencil);
```

Comment:
do you not need to reset the states somewhere after those draws?

Reply:
Yeah, I think something more general for handling state would be good. For now just caching the stuff touched in `clear_attachments` should be fine (DS state, shaders, render targets, viewports, etc.). Even then it's almost like replicating the whole state machine..

Something like this perhaps?
```rust
struct RenderTargetsGuard {
    rtvs: [*mut ID3D11RenderTargetView; MAX_RTVS],
}

impl Drop for RenderTargetsGuard {
    fn drop() {
        OMSetRenderTargets(rtvs);
    }
}

struct DxState {
    rtvs: [*mut ID3D11RenderTargetView; MAX_RTVS],
    ...
}

impl DxState {
    fn set_rtvs(self, rtvs) {
        OMSetRenderTargets(rtvs);
        self.rtvs = rtvs;
    }

    fn set_temp_rtvs(self, rtvs) -> RenderTargetsGuard {
        // set temporary
        OMSetRenderTargets(rtvs);

        // and return the state "guard" which resets state after out of scope
        RenderTargetsGuard {
            self.rtvs
        }
    }
}
```

Reply:
alternatively, you can cache all the state like Metal does and restore from the cache